blockchain/transaction_confirmers.go

```diff
@@ -301,73 +302,158 @@ func (e *EthereumClient) GetHeaderSubscriptions() map[string]HeaderEventSubscrip
 	return newMap
 }
 
-// subscribeToNewHeaders
+// subscribeToNewHeaders kicks off the primary header subscription for the test loop
 func (e *EthereumClient) subscribeToNewHeaders() error {
 	headerChannel := make(chan *SafeEVMHeader)
 	subscription, err := e.SubscribeNewHeaders(context.Background(), headerChannel)
 	if err != nil {
 		return err
 	}
-	defer subscription.Unsubscribe()
-
 	log.Info().Str("Network", e.NetworkConfig.Name).Msg("Subscribed to new block headers")
 
+	go e.headerSubscriptionLoop(subscription, headerChannel)
+	return nil
+}
+
+// headerSubscriptionLoop receives new headers, and handles subscription errors when they pop up
+func (e *EthereumClient) headerSubscriptionLoop(subscription ethereum.Subscription, headerChannel chan *SafeEVMHeader) {
+	lastHeaderNumber := uint64(0)
 	for {
 		select {
-		case err := <-subscription.Err():
-			log.Error().Err(err).Msg("Error while subscribed to new headers, restarting subscription")
-			subscription.Unsubscribe()
-
-			subscription, err = e.SubscribeNewHeaders(context.Background(), headerChannel)
+		case err := <-subscription.Err(): // Most subscription errors are temporary RPC downtime, so let's poll to resubscribe
+			log.Error().Err(err).Msg("Error while subscribed to new headers, likely RPC downtime. Attempting to resubscribe")
+			e.connectionIssueCh <- time.Now()
+			subscription = e.resubscribeLoop(headerChannel, lastHeaderNumber)
+			e.connectionRestoredCh <- time.Now()
+		case header := <-headerChannel:
+			err := e.receiveHeader(header)
 			if err != nil {
-				log.Error().Err(err).Msg("Failed to resubscribe to new headers")
-				return err
+				log.Error().Err(err).Msg("Error receiving header, possible RPC issues")
+			} else {
+				lastHeaderNumber = header.Number.Uint64()
 			}
-		case header := <-headerChannel:
-			e.receiveHeader(header)
 		case <-e.doneChan:
 			log.Debug().Str("Network", e.NetworkConfig.Name).Msg("Subscription cancelled")
 			e.Client.Close()
-			return nil
+			return
+		}
+	}
+}
+
+// resubscribeLoop polls the RPC connection until it comes back up, and then resubscribes to new headers
+func (e *EthereumClient) resubscribeLoop(headerChannel chan *SafeEVMHeader, lastHeaderNumber uint64) ethereum.Subscription {
+	rpcDegradedTime := time.Now()
+	pollInterval := time.Millisecond * 500
+	checkTicker, informTicker := time.NewTicker(pollInterval), time.NewTicker(time.Second*10)
+	reconnectAttempts, consecutiveSuccessCount, targetSuccessCount := 0, 0, 3
```

Comment:
Nit : All magic numbers in constants would help

Reply:
I can make the target count a CONST, unfortunately the other two must stay as it.